scripts/dats_jsonld_annotator/annotator.py

```diff
@@ -0,0 +1,323 @@
+import argparse
+import json
+import logging
+import pathlib as pal
+import sys
+import time
+
+import jsonschema as jss
+
+
+CONTEXT_DIR = (pal.Path(__file__).parent / "context" / "sdo").resolve()
+SCHEMA_DIR = (pal.Path(__file__).parent / "schema").resolve()
+logger = logging.getLogger("DATS annotator")
+
+
+def find_schema(parent_schema, term, json_object):
+    """
+    This function finds the appropriate JSON schema for a term in a DATS Json object. To do
+    so, it relies on the JSON schema of the supplied JSON object.
+
+    :param parent_schema: The DATS JSON schema that corresponds to the json_object parameter
+    :param term: the string term for which the schema shall be found
+    :param json_object:  the JSON object that contains the term and to which the parent_schema corresponds
+    :return: the DATS JSON schema for the term
+    """
+    ps = parent_schema["properties"]
+    if term not in ps.keys():
+        # TODO: Decide whether to raise an Exception here
+        # The only way this could happen on a valid dataset is if someone has specified additional properties
+        # We'd like to discourage this
+        logger.debug(
+            f'I cannot find {term = } in the parent schema: {parent_schema["id"]}.\n{json_object = }'
+        )
+        return None
+
+    # Set up the schema resolver
+    resolver = jss.RefResolver(base_uri=parent_schema["id"], referrer=None)
+
+    # See if "items" is in the schema entry
+    search_dict = ps[term]
+    if "items" in search_dict.keys():
+        search_dict = ps[term]["items"]
+
+    if "$ref" in search_dict.keys():
+        schema_name = search_dict["$ref"]
+    elif len(set(search_dict.keys()).intersection(["anyOf", "oneOf", "allOf"])) > 0:
```

Comment:
I wonder whether the content of this if section could be moved into a function to make the code a bit clearer. I got lost in the different if statements when reviewing the PR, LOL.

Reply:
let's keep as is with the new comments and maybe revisit later if there is another solution to finding valid schemata for a given json object